### DISCUSSION TITLE
Retrieve workflow ID for a given PR

### DISCUSSION THREAD

**Discussion by pauleustice** [community/community#24709] [Apps, API and Webhooks]
<p>Hi everyone! I need to get the artefacts from the last workflow run for a PR associated with a commit SHA.</p>
<p>The steps I envisage are:</p>
<p><strong>1.</strong> Retrieve PR by commit SHA<br>
<strong>2.</strong> Get last workflow ID for that PR<br>
<strong>3.</strong> Get artefacts by workflow ID</p>
<p>I can do <strong>1</strong> with <code>/repos/{owner}/{repo}/commits/{commit_sha}/pulls</code> and <strong>3</strong> with <code>/repos/{owner}/{repo}/actions/runs/{run_id}/artifacts</code>, but I’m missing how to do step 2.</p>
<p>I have looked at the example response data for <code>/repos/{owner}/{repo}/commits/{commit_sha}/pulls</code> but it doesn’t contain any references to workflow runs that it triggered.</p>
<p>Is there an API that allows me to retrieve a workflow ID by using a PR’s ID?</p>
<p>Thanks in advance!</p>

**Comment by jsoref** [comment]
<p>I’ve been trying to do gymnastics like this for months.</p>
<p>The GraphQL API lets you get from a <code>CheckRun</code> to a <code>WorkflowRun</code>.</p>
<p>Here’s a <code>check-run</code>:<br>
<a href="https://api.github.com/repos/jsoref/examples-testing/check-runs/3456649098" class="onebox" target="_blank" rel="noopener">https://api.github.com/repos/jsoref/examples-testing/check-runs/3456649098</a></p>
<pre><code class="lang-json">{
  "id": 3456649098,
  "name": "Spell checking",
  "node_id": "MDg6Q2hlY2tSdW4zNDU2NjQ5MDk4",
...
}
</code></pre>
<p>Taking the <code>node_id</code> and posting to <a href="https://api.github.com/graphql">https://api.github.com/graphql</a></p>
<pre><code class="lang-auto">{
  node(id: "MDg6Q2hlY2tSdW4zNDU2NjQ5MDk4")  {
    __typename
    ... on CheckRun {
      checkSuite {
        commit {
          commitUrl
        }
        workflowRun {
          url
        }
      }
      url
    }
  }
}
</code></pre>
<p>yields:</p>
<pre><code class="lang-auto">{
  "data": {
    "node": {
      "__typename": "CheckRun",
      "checkSuite": {
        "commit": {
          "commitUrl": "https://github.com/jsoref/examples-testing/commit/a826ccb09b685930af00723ee937dc11716b32b7"
        },
        "workflowRun": {
          "url": "https://github.com/jsoref/examples-testing/actions/runs/1180203998"
        }
      },
      "url": "https://github.com/jsoref/examples-testing/runs/3456649098"
    }
  }
}
</code></pre>
<p>(I’ve left the top <code>url</code> field for show, you can see it has an artifact.)</p>
<p>Taking the <code>workflowRun.url</code>:</p><aside class="onebox githubactions" data-onebox-src="https://github.com/jsoref/examples-testing/actions/runs/1180203998">
  <header class="source">

      <a href="https://github.com/jsoref/examples-testing/actions/runs/1180203998" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">


      <svg width="24" height="24" class="github-icon github-icon-failure" viewbox="0 0 24 24" aria-hidden="true"><path d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12zm8.036-4.024a.75.75 0 00-1.06 1.06L10.939 12l-2.963 2.963a.75.75 0 101.06 1.06L12 13.06l2.963 2.964a.75.75 0 001.061-1.06L13.061 12l2.963-2.964a.75.75 0 10-1.06-1.06L12 10.939 9.036 7.976z"></path></svg>

  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/jsoref/examples-testing/actions/runs/1180203998" target="_blank" rel="noopener">go back</a>
    </h4>

      <div class="github-info">
        Spell checking <span class="github-run-number">#381</span>
      </div>
  </div>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<p>
and switching to the <code>v3</code> API form (i.e. adding <code>api.</code> to the domain and <code>repos/</code> at the beginning of the path):<br>
<a href="https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998" class="onebox" target="_blank" rel="noopener">https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998</a></p>
<pre><code class="lang-json">{
  "id": 1180203998,
  "name": "Spell checking",
  "node_id": "WFR_kwLODlzVTc5GWHve",
  "head_branch": "IFooBar",
  "head_sha": "a826ccb09b685930af00723ee937dc11716b32b7",
  "run_number": 381,
  "event": "pull_request_target",
  "status": "completed",
  "conclusion": "failure",
  "workflow_id": 713920,
  "check_suite_id": 3625854331,
  "check_suite_node_id": "MDEwOkNoZWNrU3VpdGUzNjI1ODU0MzMx",
  "url": "https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998",
  "html_url": "https://github.com/jsoref/examples-testing/actions/runs/1180203998",
  "pull_requests": [
    {
      "url": "https://api.github.com/repos/jsoref/examples-testing/pulls/8",
      "id": 699280813,
      "number": 8,
      "head": {
        "ref": "IFooBar",
        "sha": "a57d9e2366f613a3a44231d4f8d79a09ec3433b3",
        "repo": {
          "id": 240964941,
          "url": "https://api.github.com/repos/jsoref/examples-testing",
          "name": "examples-testing"
        }
      },
      "base": {
        "ref": "youtube",
        "sha": "7a74808e65a9536f9b7e47995d474bbf20e3ed93",
        "repo": {
          "id": 240964941,
          "url": "https://api.github.com/repos/jsoref/examples-testing",
          "name": "examples-testing"
        }
      }
    }
  ],
  "created_at": "2021-08-29T22:23:17Z",
  "updated_at": "2021-08-29T22:23:55Z",
  "jobs_url": "https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/jobs",
  "logs_url": "https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/logs",
  "check_suite_url": "https://api.github.com/repos/jsoref/examples-testing/check-suites/3625854331",
  "artifacts_url": "https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/artifacts",
  "cancel_url": "https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/cancel",
  "rerun_url": "https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/rerun",
  "workflow_url": "https://api.github.com/repos/jsoref/examples-testing/actions/workflows/713920",
  "head_commit": {
    "id": "a826ccb09b685930af00723ee937dc11716b32b7",
    "tree_id": "4bc11cc5e7440d14c4092efba348074979944ab4",
    "message": "go back",
    "timestamp": "2021-08-29T22:23:12Z",
    "author": {
      "name": "Josh Soref",
      "email": "jsoref@users.noreply.github.com"
    },
    "committer": {
      "name": "Josh Soref",
      "email": "jsoref@users.noreply.github.com"
    }
  },
  "repository": {
    "id": 240964941,
    "node_id": "MDEwOlJlcG9zaXRvcnkyNDA5NjQ5NDE=",
    "name": "examples-testing",
    "full_name": "jsoref/examples-testing",
    "private": false,
    "owner": {
      "login": "jsoref",
      "id": 2119212,
      "node_id": "MDQ6VXNlcjIxMTkyMTI=",
      "avatar_url": "https://avatars.githubusercontent.com/u/2119212?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/jsoref",
      "html_url": "https://github.com/jsoref",
      "followers_url": "https://api.github.com/users/jsoref/followers",
      "following_url": "https://api.github.com/users/jsoref/following{/other_user}",
      "gists_url": "https://api.github.com/users/jsoref/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/jsoref/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/jsoref/subscriptions",
      "organizations_url": "https://api.github.com/users/jsoref/orgs",
      "repos_url": "https://api.github.com/users/jsoref/repos",
      "events_url": "https://api.github.com/users/jsoref/events{/privacy}",
      "received_events_url": "https://api.github.com/users/jsoref/received_events",
      "type": "User",
      "site_admin": false
    },
    "html_url": "https://github.com/jsoref/examples-testing",
    "description": "This repository will be trashed periodically -- it's for testing / debugging the examples repository",
    "fork": true,
    "url": "https://api.github.com/repos/jsoref/examples-testing",
    "forks_url": "https://api.github.com/repos/jsoref/examples-testing/forks",
    "keys_url": "https://api.github.com/repos/jsoref/examples-testing/keys{/key_id}",
    "collaborators_url": "https://api.github.com/repos/jsoref/examples-testing/collaborators{/collaborator}",
    "teams_url": "https://api.github.com/repos/jsoref/examples-testing/teams",
    "hooks_url": "https://api.github.com/repos/jsoref/examples-testing/hooks",
    "issue_events_url": "https://api.github.com/repos/jsoref/examples-testing/issues/events{/number}",
    "events_url": "https://api.github.com/repos/jsoref/examples-testing/events",
    "assignees_url": "https://api.github.com/repos/jsoref/examples-testing/assignees{/user}",
    "branches_url": "https://api.github.com/repos/jsoref/examples-testing/branches{/branch}",
    "tags_url": "https://api.github.com/repos/jsoref/examples-testing/tags",
    "blobs_url": "https://api.github.com/repos/jsoref/examples-testing/git/blobs{/sha}",
    "git_tags_url": "https://api.github.com/repos/jsoref/examples-testing/git/tags{/sha}",
    "git_refs_url": "https://api.github.com/repos/jsoref/examples-testing/git/refs{/sha}",
    "trees_url": "https://api.github.com/repos/jsoref/examples-testing/git/trees{/sha}",
    "statuses_url": "https://api.github.com/repos/jsoref/examples-testing/statuses/{sha}",
    "languages_url": "https://api.github.com/repos/jsoref/examples-testing/languages",
    "stargazers_url": "https://api.github.com/repos/jsoref/examples-testing/stargazers",
    "contributors_url": "https://api.github.com/repos/jsoref/examples-testing/contributors",
    "subscribers_url": "https://api.github.com/repos/jsoref/examples-testing/subscribers",
    "subscription_url": "https://api.github.com/repos/jsoref/examples-testing/subscription",
    "commits_url": "https://api.github.com/repos/jsoref/examples-testing/commits{/sha}",
    "git_commits_url": "https://api.github.com/repos/jsoref/examples-testing/git/commits{/sha}",
    "comments_url": "https://api.github.com/repos/jsoref/examples-testing/comments{/number}",
    "issue_comment_url": "https://api.github.com/repos/jsoref/examples-testing/issues/comments{/number}",
    "contents_url": "https://api.github.com/repos/jsoref/examples-testing/contents/{+path}",
    "compare_url": "https://api.github.com/repos/jsoref/examples-testing/compare/{base}...{head}",
    "merges_url": "https://api.github.com/repos/jsoref/examples-testing/merges",
    "archive_url": "https://api.github.com/repos/jsoref/examples-testing/{archive_format}{/ref}",
    "downloads_url": "https://api.github.com/repos/jsoref/examples-testing/downloads",
    "issues_url": "https://api.github.com/repos/jsoref/examples-testing/issues{/number}",
    "pulls_url": "https://api.github.com/repos/jsoref/examples-testing/pulls{/number}",
    "milestones_url": "https://api.github.com/repos/jsoref/examples-testing/milestones{/number}",
    "notifications_url": "https://api.github.com/repos/jsoref/examples-testing/notifications{?since,all,participating}",
    "labels_url": "https://api.github.com/repos/jsoref/examples-testing/labels{/name}",
    "releases_url": "https://api.github.com/repos/jsoref/examples-testing/releases{/id}",
    "deployments_url": "https://api.github.com/repos/jsoref/examples-testing/deployments"
  },
  "head_repository": {
    "id": 240964941,
    "node_id": "MDEwOlJlcG9zaXRvcnkyNDA5NjQ5NDE=",
    "name": "examples-testing",
    "full_name": "jsoref/examples-testing",
    "private": false,
    "owner": {
      "login": "jsoref",
      "id": 2119212,
      "node_id": "MDQ6VXNlcjIxMTkyMTI=",
      "avatar_url": "https://avatars.githubusercontent.com/u/2119212?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/jsoref",
      "html_url": "https://github.com/jsoref",
      "followers_url": "https://api.github.com/users/jsoref/followers",
      "following_url": "https://api.github.com/users/jsoref/following{/other_user}",
      "gists_url": "https://api.github.com/users/jsoref/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/jsoref/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/jsoref/subscriptions",
      "organizations_url": "https://api.github.com/users/jsoref/orgs",
      "repos_url": "https://api.github.com/users/jsoref/repos",
      "events_url": "https://api.github.com/users/jsoref/events{/privacy}",
      "received_events_url": "https://api.github.com/users/jsoref/received_events",
      "type": "User",
      "site_admin": false
    },
    "html_url": "https://github.com/jsoref/examples-testing",
    "description": "This repository will be trashed periodically -- it's for testing / debugging the examples repository",
    "fork": true,
    "url": "https://api.github.com/repos/jsoref/examples-testing",
    "forks_url": "https://api.github.com/repos/jsoref/examples-testing/forks",
    "keys_url": "https://api.github.com/repos/jsoref/examples-testing/keys{/key_id}",
    "collaborators_url": "https://api.github.com/repos/jsoref/examples-testing/collaborators{/collaborator}",
    "teams_url": "https://api.github.com/repos/jsoref/examples-testing/teams",
    "hooks_url": "https://api.github.com/repos/jsoref/examples-testing/hooks",
    "issue_events_url": "https://api.github.com/repos/jsoref/examples-testing/issues/events{/number}",
    "events_url": "https://api.github.com/repos/jsoref/examples-testing/events",
    "assignees_url": "https://api.github.com/repos/jsoref/examples-testing/assignees{/user}",
    "branches_url": "https://api.github.com/repos/jsoref/examples-testing/branches{/branch}",
    "tags_url": "https://api.github.com/repos/jsoref/examples-testing/tags",
    "blobs_url": "https://api.github.com/repos/jsoref/examples-testing/git/blobs{/sha}",
    "git_tags_url": "https://api.github.com/repos/jsoref/examples-testing/git/tags{/sha}",
    "git_refs_url": "https://api.github.com/repos/jsoref/examples-testing/git/refs{/sha}",
    "trees_url": "https://api.github.com/repos/jsoref/examples-testing/git/trees{/sha}",
    "statuses_url": "https://api.github.com/repos/jsoref/examples-testing/statuses/{sha}",
    "languages_url": "https://api.github.com/repos/jsoref/examples-testing/languages",
    "stargazers_url": "https://api.github.com/repos/jsoref/examples-testing/stargazers",
    "contributors_url": "https://api.github.com/repos/jsoref/examples-testing/contributors",
    "subscribers_url": "https://api.github.com/repos/jsoref/examples-testing/subscribers",
    "subscription_url": "https://api.github.com/repos/jsoref/examples-testing/subscription",
    "commits_url": "https://api.github.com/repos/jsoref/examples-testing/commits{/sha}",
    "git_commits_url": "https://api.github.com/repos/jsoref/examples-testing/git/commits{/sha}",
    "comments_url": "https://api.github.com/repos/jsoref/examples-testing/comments{/number}",
    "issue_comment_url": "https://api.github.com/repos/jsoref/examples-testing/issues/comments{/number}",
    "contents_url": "https://api.github.com/repos/jsoref/examples-testing/contents/{+path}",
    "compare_url": "https://api.github.com/repos/jsoref/examples-testing/compare/{base}...{head}",
    "merges_url": "https://api.github.com/repos/jsoref/examples-testing/merges",
    "archive_url": "https://api.github.com/repos/jsoref/examples-testing/{archive_format}{/ref}",
    "downloads_url": "https://api.github.com/repos/jsoref/examples-testing/downloads",
    "issues_url": "https://api.github.com/repos/jsoref/examples-testing/issues{/number}",
    "pulls_url": "https://api.github.com/repos/jsoref/examples-testing/pulls{/number}",
    "milestones_url": "https://api.github.com/repos/jsoref/examples-testing/milestones{/number}",
    "notifications_url": "https://api.github.com/repos/jsoref/examples-testing/notifications{?since,all,participating}",
    "labels_url": "https://api.github.com/repos/jsoref/examples-testing/labels{/name}",
    "releases_url": "https://api.github.com/repos/jsoref/examples-testing/releases{/id}",
    "deployments_url": "https://api.github.com/repos/jsoref/examples-testing/deployments"
  }
}
</code></pre>
<p>which includes:</p>
<pre><code class="lang-auto">  "artifacts_url": "https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/artifacts",
</code></pre>
<p>And:<br>
<a href="https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/artifacts" class="onebox" target="_blank" rel="noopener">https://api.github.com/repos/jsoref/examples-testing/actions/runs/1180203998/artifacts</a><br>
Has an artifact.</p>
<p>I personally needed the <code>WorkflowRun</code>'s <code>event</code> (<code>pull_request_target</code>) and the <code>workflow_url</code> (to get <code>path</code>, <code>name</code>, and <code>state</code>).</p>
<p>Fwiw, afaict for my case, I actually must switch back and forth between the <code>v3</code> and <code>v4</code> api, as the <code>v4</code> api content for the <code>WorkflowRun</code> is really bare. In your case, it might be possible to work solely with the <code>v4</code> api.</p>

**Comment by pauleustice** [comment]
<p>Thanks for responding <a class="mention" href="/jsoref">@jsoref</a>, ‘gymnastics’ is definitely the right word! I’m afraid I still can’t see how to find the missing link between a PR and workflow runs though, as the PR doesn’t return check runs from what I can tell. I’ve just been delving into the GraphQL schema and can’t find a link there at all.</p>
<p>It’s frustrating! The only thing I can think of at the moment is to loop over all workflow runs and find a way to filter them by the related commit SHA or PR number. This is obviously hugely wasteful when the GraphQL layer should (to my mind) be able to do the legwork for us.</p>

**Comment by pauleustice** [comment, marked as answer]
<p>I spoke to GH Support who confirmed that the only way to do this is to filter the workflow runs.</p>
<p>This is the approach:</p>
<ol>
<li>
<p>Use a given SHA to retrieve PR:<br>
/repos/{owner}/{repo}/commits/{commit_sha}/pulls</p>
</li>
<li>
<p>Get all relevant workflows<br>
/repos/{owner}/{repo}/actions/runs</p>
</li>
</ol>
<p>Query params:<br>
<code>branch</code>: <code>pr.head.ref</code> (where <code>pr</code> is the PR data object from step 1)<br>
<code>event</code>: <code>"pull_request"</code></p>
<ol start="3">
<li>
<p>Filter workflow runs for that particular commit using the commit SHA<br>
where <code>workflow_run.head_sha</code> === <code>commit_sha</code></p>
</li>
<li>
<p>Retrieve artefacts by using the workflow run ID from step 3<br>
<code>/repos/{owner}/{repo}/actions/runs/{run_id}/artifacts</code></p>
</li>
</ol>

**Reply by champak-7span** [reply]
I have tried to with the query param but I did not get any results.  if you have any got any solution then please put it here .. thanks

**Reply by volatilemolotov** [reply]
use `head_sha` query param
